File: sdk/android/src/test/kotlin/com/shieldnet360/access/ContractTest.kt
/*
 * ContractTest.kt — compile-time conformance check.
 *
 * The SDK ships an interface only; concrete `OkHttp`-backed clients live
 * in the host application. This test asserts that the interface is
 * implementable: if a method signature changes in a breaking way, the
 * mock below stops compiling and the test target fails to build. There
 * are intentionally **no network round-trip tests** here.
 */
package com.shieldnet360.access

import java.time.Instant
import kotlinx.coroutines.runBlocking
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/** Stub conformance — exercises every method on the interface. */
private class MockAccessSDKClient : AccessSDKClient {
    override suspend fun createRequest(
        resource: String,
        role: String?,
        justification: String?,
    ): AccessRequest = AccessRequest(
        id = "req_test",
        workspaceId = "ws_test",
        requesterUserId = "user_test",
        connectorId = "conn_test",
        resourceExternalId = resource,
        role = role,
        justification = justification,
        state = AccessRequestState.REQUESTED,
        createdAt = Instant.EPOCH,
    )

    override suspend fun listRequests(
        state: AccessRequestState?,
        requester: String?,
        resource: String?,
    ): List<AccessRequest> = emptyList()

    override suspend fun approveRequest(id: String): AccessRequest = AccessRequest(
        id = id,
        workspaceId = "ws_test",
        requesterUserId = "user_test",
        connectorId = "conn_test",
        resourceExternalId = "res_test",
        state = AccessRequestState.APPROVED,
        createdAt = Instant.EPOCH,
    )

    override suspend fun denyRequest(id: String, reason: String): AccessRequest = AccessRequest(
        id = id,
        workspaceId = "ws_test",
        requesterUserId = "user_test",
        connectorId = "conn_test",
        resourceExternalId = "res_test",
        state = AccessRequestState.DENIED,
        createdAt = Instant.EPOCH,
    )

    override suspend fun cancelRequest(id: String): AccessRequest = AccessRequest(
        id = id,
        workspaceId = "ws_test",
        requesterUserId = "user_test",
        connectorId = "conn_test",
        resourceExternalId = "res_test",
        state = AccessRequestState.CANCELLED,
        createdAt = Instant.EPOCH,
    )

    override suspend fun listGrants(userId: String?, connectorId: String?): List<AccessGrant> = emptyList()

    override suspend fun explainPolicy(policyId: String): PolicyExplanation =
        PolicyExplanation(policyId = policyId, summary = "mock")

    override suspend fun suggestResources(): List<Suggestion> = emptyList()
}

class ContractTest {
    @Test
    fun `mock satisfies AccessSDKClient and can be exercised end-to-end`() = runBlocking {
        val client: AccessSDKClient = MockAccessSDKClient()

        val created = client.createRequest(
            resource = "res_test",
            role = "viewer",
            justification = "ci",
        )
        assertEquals(AccessRequestState.REQUESTED, created.state)
        assertEquals("res_test", created.resourceExternalId)
        assertEquals("conn_test", created.connectorId)

        val list = client.listRequests()
        assertTrue(list.isEmpty())

        val approved = client.approveRequest("req_1")
        assertEquals(AccessRequestState.APPROVED, approved.state)

        val denied = client.denyRequest("req_2", "policy-violation")
        assertEquals(AccessRequestState.DENIED, denied.state)

        val cancelled = client.cancelRequest("req_3")
        assertEquals(AccessRequestState.CANCELLED, cancelled.state)

        val grants = client.listGrants()
        assertTrue(grants.isEmpty())

        val explanation = client.explainPolicy("pol_1")
        assertEquals("pol_1", explanation.policyId)

        val suggestions = client.suggestResources()
        assertTrue(suggestions.isEmpty())
    }

    @Test
    fun `AccessSDKException carries structured status codes`() {
        val ex = AccessSDKException.Http(statusCode = 401, body = "{\"error\":{\"code\":\"unauthorized\"}}")
        assertEquals(401, ex.statusCode)
        assertTrue(ex.message!!.contains("HTTP 401"))
    }

    @Test
    fun `AccessRequestState maps to and from lowercase wire values`() {
        // Wire values match the Go-side constants in
        // internal/services/access/request_state_machine.go.
        assertEquals("requested", AccessRequestState.REQUESTED.value)
        assertEquals("provision_failed", AccessRequestState.PROVISION_FAILED.value)
        assertEquals("expired", AccessRequestState.EXPIRED.value)

        assertEquals(AccessRequestState.APPROVED, AccessRequestState.fromWire("approved"))
        assertEquals(AccessRequestState.PROVISION_FAILED, AccessRequestState.fromWire("provision_failed"))
        assertEquals(AccessRequestState.EXPIRED, AccessRequestState.fromWire("expired"))

        val unknown = try {
            AccessRequestState.fromWire("unknown_state")
            null
        } catch (e: IllegalArgumentException) {
            e
        }
        assertTrue(unknown != null, "fromWire must throw on unknown values")
    }

    @Test
    fun `AccessRequestRiskScore maps to and from lowercase wire values`() {
        // Wire values match the Go-side RequestRiskLow / Medium / High
        // constants in internal/models/access_request.go.
        assertEquals("low", AccessRequestRiskScore.LOW.value)
        assertEquals("medium", AccessRequestRiskScore.MEDIUM.value)
        assertEquals("high", AccessRequestRiskScore.HIGH.value)

        assertEquals(AccessRequestRiskScore.LOW, AccessRequestRiskScore.fromWire("low"))
        assertEquals(AccessRequestRiskScore.HIGH, AccessRequestRiskScore.fromWire("high"))
    }
}

File: sdk/desktop/src/__tests__/contract.test.ts
/**
 * contract.test.ts — compile-time conformance check.
 *
 * The desktop extension ships type definitions only; concrete
 * implementations live in the host Electron application. This file is a
 * pure type-check: it constructs a mock object that satisfies the
 * `AccessIPC` interface. If a method signature changes in a breaking way,
 * `tsc --noEmit` fails and the test target fails to build. There are
 * intentionally **no runtime IPC round-trip tests** here.
 */

import {
  AccessGrant,
  AccessIPC,
  AccessIPCChannel,
  AccessIPCError,
  AccessRequest,
  AccessRequestPayload,
  AccessRequestResponse,
  AccessRequestListFilter,
  AccessRequestListResponse,
  AIQueryResponse,
  GrantListResponse,
  PolicyExplanation,
  PolicyQueryPayload,
  PolicyQueryResponse,
  Suggestion,
} from '../access-ipc';

function fakeRequest(state: AccessRequest['state'], id = 'req_test'): AccessRequest {
  return {
    id,
    workspaceId: 'ws_test',
    requesterUserId: 'user_test',
    resourceExternalId: 'res_test',
    state,
    createdAt: new Date(0).toISOString(),
  };
}

const mockGrant: AccessGrant = {
  id: 'grant_test',
  workspaceId: 'ws_test',
  userId: 'user_test',
  connectorId: 'conn_test',
  resourceExternalId: 'res_test',
  grantedAt: new Date(0).toISOString(),
};

const mockExplanation: PolicyExplanation = {
  policyId: 'pol_1',
  summary: 'mock',
  rationale: [],
  affectedResources: [],
};

const mockSuggestion: Suggestion = {
  id: 'sug_1',
  resourceExternalId: 'res_test',
  displayName: 'Mock Resource',
  reason: 'frequently accessed by your team',
};

const mockClient: AccessIPC = {
  requestAccess: {
    async create(payload: AccessRequestPayload): Promise<AccessRequestResponse> {
      return { request: fakeRequest('requested') };
    },
    async list(_filter?: AccessRequestListFilter): Promise<AccessRequestListResponse> {
      return { requests: [], nextCursor: null };
    },
    async approve(id: string): Promise<AccessRequestResponse> {
      return { request: fakeRequest('approved', id) };
    },
    async deny(id: string, _payload): Promise<AccessRequestResponse> {
      return { request: fakeRequest('denied', id) };
    },
    async cancel(id: string): Promise<AccessRequestResponse> {
      return { request: fakeRequest('cancelled', id) };
    },
  },

  async listGrants(): Promise<GrantListResponse> {
    return { grants: [mockGrant], nextCursor: null };
  },

  async queryPolicy(_payload: PolicyQueryPayload): Promise<PolicyQueryResponse> {
    return { explanation: mockExplanation };
  },

  async askAI(): Promise<AIQueryResponse> {
    return { suggestions: [mockSuggestion] };
  },
};

// Force usage so tsc doesn't tree-shake the assertions.
const _channelCheck: typeof AccessIPCChannel.RequestAccessCreate = 'access:requestAccess.create';

async function exerciseContract(client: AccessIPC): Promise<void> {
  const created = await client.requestAccess.create({
    resourceExternalId: 'res_test',
    role: 'viewer',
    justification: 'ci',
  });
  if (created.request.state !== 'requested') {
    throw new Error('expected requested');
  }

  const list = await client.requestAccess.list();
  if (list.requests.length !== 0) {
    throw new Error('expected empty list');
  }

  const approved = await client.requestAccess.approve('req_1');
  if (approved.request.state !== 'approved') {
    throw new Error('expected approved');
  }

  const denied = await client.requestAccess.deny('req_2', { reason: 'policy-violation' });
  if (denied.request.state !== 'denied') {
    throw new Error('expected denied');
  }

  const cancelled = await client.requestAccess.cancel('req_3');
  if (cancelled.request.state !== 'cancelled') {
    throw new Error('expected cancelled');
  }

  const grants = await client.listGrants();
  if (grants.grants.length !== 1) {
    throw new Error('expected one grant');
  }

  const explanation = await client.queryPolicy({ policyId: 'pol_1' });
  if (explanation.explanation.policyId !== 'pol_1') {
    throw new Error('expected pol_1');
  }

  const ai = await client.askAI();
  if (ai.suggestions.length !== 1) {
    throw new Error('expected one suggestion');
  }

  const err = new AccessIPCError('http', 'forbidden', { statusCode: 403 });
  if (err.kind !== 'http' || err.statusCode !== 403) {
    throw new Error('expected http kind / 403');
  }

  void _channelCheck;
}

// Top-level expression so the type-checker actually evaluates the
// assignments above and so the mock isn't tree-shaken.
void exerciseContract(mockClient);
